helix-rest/src/main/java/org/apache/helix/rest/server/resources/helix/ClusterAccessor.java

```diff
@@ -442,6 +442,29 @@ public Response getClusterTopology(@PathParam("clusterId") String clusterId) thr
     return OK(objectMapper.writeValueAsString(clusterTopology));
   }
 
+  @ResponseMetered(name = HttpConstants.READ_REQUEST)
+  @Timed(name = HttpConstants.READ_REQUEST)
+  @GET
+  @Path("{clusterId}/topologymap")
+  public Response getClusterTopologyMap(@PathParam("clusterId") String clusterId) {
+    if (!doesClusterExist(clusterId)) {
```

Comment:
Since this will add one more zk read, and `admin.getClusterTopology(clusterId)` actually reads the cluster, I am not sure how heavy this call is. I assume it is mostly used during deployment? 

I wonder if we can use the result in `admin.getClusterTopology`: if the cluster doesn't exist, an exception(`HelixException/NoClusterExistException`) is thrown. Rest code here catches it and returns 404. Then it saves one zk read.

Reply:
I tried a bit. So here're the cases.
There are a few exceptions. 
cluster not set up.
fault zone is not set.
instance configs are not set.
topology is not set or wrongly set.
How are we going to categorize them? I think they're all client side exceptions. What response we would like to return?

Reply:
I realized that this call is not a lightweight call. It has many calls to zk to check all required below paths exist. I think this is pretty heavy. And even the cluster exists at the usual case, this check makes the rest api call pretty slow to respond, which I don't think is expected, right? And actually `isClusterSetup` is called in `getClusterTopology` -> `getClusterConfig()` -> `getConfigZnRecord` -- multiple duplicate calls to zk just to check `isClusterSetup`.  So the following `HelixException` catch is good enough. No need to do the pre-check `if (!doesClusterExist(clusterId)) {`

```
    requiredPaths.add(PropertyPathBuilder.idealState(clusterName));
    requiredPaths.add(PropertyPathBuilder.clusterConfig(clusterName));
    requiredPaths.add(PropertyPathBuilder.instanceConfig(clusterName));
    requiredPaths.add(PropertyPathBuilder.resourceConfig(clusterName));
    requiredPaths.add(PropertyPathBuilder.propertyStore(clusterName));
    requiredPaths.add(PropertyPathBuilder.liveInstance(clusterName));
    requiredPaths.add(PropertyPathBuilder.instance(clusterName));
    requiredPaths.add(PropertyPathBuilder.externalView(clusterName));
    requiredPaths.add(PropertyPathBuilder.controller(clusterName));
    requiredPaths.add(PropertyPathBuilder.stateModelDef(clusterName));
    requiredPaths.add(PropertyPathBuilder.controllerMessage(clusterName));
    requiredPaths.add(PropertyPathBuilder.controllerError(clusterName));
    requiredPaths.add(PropertyPathBuilder.controllerStatusUpdate(clusterName));
    requiredPaths.add(PropertyPathBuilder.controllerHistory(clusterName));
```